src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs

```diff
@@ -211,24 +209,90 @@ private static void EmitCallsToMetadataProvidersForParameters(this Endpoint endp
                 break;
             }
 
+            if (parameter.IsEndpointParameterMetadataProvider)
+            {
+                codeWriter.WriteLine($$"""var {{parameter.SymbolName}}_ParameterInfo = parameterInfos.Single(p => p.Name == "{{parameter.SymbolName}}");""");
+                codeWriter.WriteLine($"PopulateMetadataForParameter<{parameterType.ToDisplayString(EmitterConstants.DisplayFormat)}>({parameter.SymbolName}_ParameterInfo, options!.EndpointBuilder);");
+            }
+
             if (parameter.IsEndpointMetadataProvider)
             {
-                codeWriter.WriteLine($"PopulateMetadataForEndpoint<{parameterType.ToDisplayString(EmitterConstants.DisplayFormat)}>(methodInfo, options.EndpointBuilder);");
+                codeWriter.WriteLine($"PopulateMetadataForEndpoint<{parameterType.ToDisplayString(EmitterConstants.DisplayFormat)}>(methodInfo, options!.EndpointBuilder);");
             }
 
-            if (parameter.IsEndpointParameterMetadataProvider)
+        }
+    }
+
+    public static void EmitFormAcceptsMetadata(this Endpoint endpoint, CodeWriter codeWriter)
+    {
+        var hasFormFiles = endpoint.Parameters.Any(p => p.IsFormFile);
+
+        if (hasFormFiles)
+        {
+            codeWriter.WriteLine("options.EndpointBuilder.Metadata.Add(new GeneratedAcceptsMetadata(contentTypes: GeneratedMetadataConstants.FormFileContentType));");
+        }
+        else
+        {
+            codeWriter.WriteLine("options.EndpointBuilder.Metadata.Add(new GeneratedAcceptsMetadata(contentTypes: GeneratedMetadataConstants.FormContentType));");
+        }
+    }
+
+    public static void EmitJsonAcceptsMetadata(this Endpoint endpoint, CodeWriter codeWriter)
+    {
+        codeWriter.WriteLine("var serviceProvider = options?.ServiceProvider ?? options?.EndpointBuilder?.ApplicationServices;");
+        codeWriter.WriteLine($"var serviceProviderIsService = serviceProvider?.GetService<IServiceProviderIsService>();");
+
+        if (endpoint.Parameters.SingleOrDefault(p => p.Source == EndpointParameterSource.JsonBody) is { } explicitBodyParameter)
```

Comment:
Use a for-each over `endpoint.Parameters` with if-checks inside instead of enumerating over the parameters twice?

Reply:
Reduced some of the LINQ usage within the loops but there are still ultimately two loops - one over the parameters the first time to try and spot an explicit `[FromBody]` and another to loop over the implicit from body or service parameters if that fails.

I had a bit of a noodle on avoiding that second loop but it would mean the emitted code is pretty ugly (which could only be avoided with another loop). So I think this now strikes the balance of compile time efficiency and emitted code aesthetics.